frontend/tests/Utils.php

```diff
@@ -433,6 +433,31 @@ public static function runAggregateFeedback(): void {
         );
     }
 
+    public static function runGenerateContestCertificates(): void {
+        // Ensure all suggestions are written to the database before invoking
+        // the external script.
+        self::commit();
+        $host_arg = '';
+        $host_chunks = explode(':', OMEGAUP_DB_HOST, 2);
+        if (count($host_chunks) == 2) {
+            [$hostname, $port] = $host_chunks;
+            $host_arg .= ' --host ' . escapeshellarg($hostname);
+            $host_arg .= ' --port ' . escapeshellarg($port);
+        } else {
+            [$hostname] = $host_chunks;
+            $host_arg .= ' --host ' . escapeshellarg($hostname);
+        }
+        self::shellExec(
+            ('python3 ' .
+            dirname(__DIR__, 2) . '/stuff/pipelines/client_contest.py' .
+            $host_arg .
```

Comment:
```suggestion
            $host_arg .
            ' --database=' . OMEGAUP_DB_NAME .
```

esto debería arreglar el fallo. la base de datos no siempre se llama `omegaup-test`, resulta.

Reply:
pero entonces, la base de datos de prueba se llama en este caso omegaup?

Reply:
se llama lo que sea que tiene `OMEGAUP_DB_NAME`. esa es la variable que usa el resto del sistema, así que no hay ninguna razón para hardcodear ningún valor aquí.

Reply:
ohohoh okay, lo que pasa es que al ponerlo asi como me dijiste, me aparece que la url es invalida, y me parece curioso porque no es un valor que yo le asigne, no se bien como este funcionando esa parte del codigo

`Exception: {'status': 'error', 'error': 'Url de scoreboard inválido', 'errorname': 'invalidScoreboardUrl', 'errorcode': 403, 'header': 'HTTP/1.1 403 FORBIDDEN', '_id': '637fd15def5e9457132241'}`

Reply:
Hola, de nuevo empece a debuggear, y aunque puse OMEGAUP_DB_NAME en --database, se siguen conectando a bd distintas

Reply:
a ver intenta pasarle los cuatro parámetros que se usan en https://github.com/omegaup/omegaup/blob/dc98fbb3561f78156d10b801bf46b10749a41494/frontend/server/src/MySQLConnection.php#L83 para confirmar.

idealmente que se parezca lo más posible a https://github.com/omegaup/omegaup/blob/dc98fbb3561f78156d10b801bf46b10749a41494/frontend/tests/Utils.php#L436-L459 (bonus points si se hace un refactor de _todas_ esas funciones para que nada más se construya esa cadena de parámetros una vez, desde `--verbose` hasta `--password`).

Reply:
Siguiendo el formato de las funciones que ya existen puse lo siguiente (lo extra que agregue fue el api token y la url):
```
public static function runGenerateContestCertificates(): void {
        // Ensure all suggestions are written to the database before invoking
        // the external script.
        self::commit();
        $host_arg = '';
        $host_chunks = explode(':', OMEGAUP_DB_HOST, 2);
        if (count($host_chunks) == 2) {
            [$hostname, $port] = $host_chunks;
            $host_arg .= ' --host ' . escapeshellarg($hostname);
            $host_arg .= ' --port ' . escapeshellarg($port);
        } else {
            [$hostname] = $host_chunks;
            $host_arg .= ' --host ' . escapeshellarg($hostname);
        }
        self::shellExec(
            ('python3 ' .
            dirname(__DIR__, 2) . '/stuff/pipelines/client_contest.py' .
             ' --verbose ' .
             ' --logfile ' . escapeshellarg(OMEGAUP_LOG_FILE) .
             $host_arg .
             ' --user ' . escapeshellarg(OMEGAUP_DB_USER) .
             ' --database ' . escapeshellarg(OMEGAUP_DB_NAME) .
             ' --password ' . escapeshellarg(OMEGAUP_DB_PASS).
             ' --api-token ' . '92d8c5a0eceef3c05f4149fc04b62bb2cd50d9c6' .
             ' --url ' . escapeshellarg(OMEGAUP_URL))
        );
    }
```
Pero aun asi me marca el siguiente error:
```
There was 1 error:

1) CertificatesTest::testConnectionRabbitMQ
Exception: Failed to run `python3 /opt/omegaup/stuff/pipelines/client_contest.py --verbose  --logfile '/opt/omegaup/frontend/tests/runfiles/shard-0/omegaup.log' --host 'mysql' --port '13306' --user 'omegaup' --database 'omegaup-test' --password 'omegaup' --api-token 92d8c5a0eceef3c05f4149fc04b62bb2cd50d9c6 --url 'http://localhost:8001/'`: status=1
stdout=
stderr=Traceback (most recent call last):
  File "/opt/omegaup/stuff/pipelines/client_contest.py", line 72, in <module>
    main()
  File "/opt/omegaup/stuff/pipelines/client_contest.py", line 61, in main
    rabbitmq_client.receive_messages(queue='contest',
  File "/opt/omegaup/stuff/pipelines/rabbitmq_client.py", line 38, in receive_messages
    channel.start_consuming()
  File "/opt/omegaup/stuff/venv/lib/python3.10/site-packages/pika/adapters/blocking_connection.py", line 1883, in start_consuming
    self._process_data_events(time_limit=None)
  File "/opt/omegaup/stuff/venv/lib/python3.10/site-packages/pika/adapters/blocking_connection.py", line 2044, in _process_data_events
    self.connection.process_data_events(time_limit=time_limit)
  File "/opt/omegaup/stuff/venv/lib/python3.10/site-packages/pika/adapters/blocking_connection.py", line 851, in process_data_events
    self._dispatch_channel_events()
  File "/opt/omegaup/stuff/venv/lib/python3.10/site-packages/pika/adapters/blocking_connection.py", line 567, in _dispatch_channel_events
    impl_channel._get_cookie()._dispatch_events()
  File "/opt/omegaup/stuff/venv/lib/python3.10/site-packages/pika/adapters/blocking_connection.py", line 1510, in _dispatch_events
    consumer_info.on_message_callback(self, evt.method,
  File "/opt/omegaup/stuff/pipelines/contest_callback.py", line 57, in __call__
    scoreboard = self.client.contest.scoreboard(
  File "/opt/omegaup/stuff/venv/lib/python3.10/site-packages/omegaup/api.py", line 15007, in scoreboard
    return _Scoreboard(**self._client.query('/api/contest/scoreboard/',
  File "/opt/omegaup/stuff/venv/lib/python3.10/site-packages/omegaup/api.py", line 22127, in query
    raise Exception(response)
Exception: {'status': 'error', 'error': 'Concurso no encontrado', 'errorname': 'contestNotFound', 'errorcode': 404, 'header': 'HTTP/1.1 404 NOT FOUND', '_id': '6385800141f1f259358917'}


/opt/omegaup/frontend/tests/Utils.php:362
/opt/omegaup/frontend/tests/Utils.php:460
/opt/omegaup/frontend/tests/controllers/CertificatesTest.php:99
```

Reply:
ah! ok, eso es progreso: al menos ya no se queja de la base de datos. y ahora está intentando mandar a llamar un API que no existe: en las pruebas de unidad no está corriendo ningún servidor! entonces ese código no va a funcionar así. quizás lo que conviene hacer es que phpunit guarde el JSON del scoreboard en un archivo temporal y le diga al cronjob dónde leerlo.

Reply:
Okay, lo checo, aunque entonces lo que entiendo es que en donde le paso el JSON:

```
$messageArray = [
            'certificate_cutoff' => $contestExtraInformation['certificate_cutoff'],
            'alias' => $contestExtraInformation['alias'],
            'scoreboard_url' => $contestExtraInformation['scoreboard_url'],
            'contest_id' => $contestExtraInformation['contest_id']
        ];
$messageJSON = json_encode($messageArray);
        $message = new \PhpAmqpLib\Message\AMQPMessage($messageJSON);
```
ahora el JSON lo voy a tener que guardar en un archivo temporal o solo el JSON del scoreboard, como que no entendi muy bien

Reply:
En caso de crear un archivo temporal, en donde son creados este tipo de archivos?

Reply:
ah estamos hablando de JSONs distintos: ese que mencionas hay que dejarlo tal cual (ese es el payload con el que le mandamos decir al generador que procese ese concurso). el otro es el JSON que resulta de mandar llamar a la API de scoreboards.

aunque ahora que lo pienso: creo que no tiene sentido para nada que tengamos al frontend que "mande llamar" al generador con información incompleta, y luego que el generador le tenga que pedir esa información al frontend!

_entonces_ creo que lo mejor es ni si quiera hacer ese JSON temporal que había sugerido. lo mejor sería agregarle el resultado del scoreboard al JSON ese que ya le estamos mandando, para que el cronjob no tenga que mandar llamar ni un sólo API.

Reply:
seria `omegaup.api `dentro del archivo de `client_contest.py` solamente?

Reply:
entonces lo que hay que hacer, en otro cambio:

* En `stuff/pipelines/database/contest.py` agregar información del scoreboard en `ContestCertificate`.
* Mover la llamada de `self.client.contest.scoreboard(...)` de `stuff/pipelines/contest_callback.py` a `stuff/pipelines/producer_contest.py`, de tal manera que en `stuff/pipelines/contest_callback.py` no se necesite llamar al API.
* Eliminar todas las referencias de `omegaup.api` de `stuff/pipelines/client_contest.py` y `stuff/pipelines/contest_callback.py`. 

Ya con eso, desde acá desde PHP ya se debería pasar la información del scoreboard de antemano.

Reply:
hice estos cambios en [https://github.com/omegaup/omegaup/pull/6863](url)